internal/pkg/service/appsproxy/proxy/transport/transport.go

```diff
@@ -0,0 +1,138 @@
+const TLSHandshakeTimeout = 5 * time.Second
+
+// ResponseHeaderTimeout specifies the default amount of time to wait for a server's response headers.
+const ResponseHeaderTimeout = 15 * time.Second
+
+// MaxIdleConnections specifies the default maximum number of open connections at all.
+const MaxIdleConnections = 128
+
+// MaxConnectionsPerHost specifies the default maximum number of open connections to a host.
+const MaxConnectionsPerHost = 32
+
+// HTTP2ReadIdleTimeout is the timeout after which a health check using ping frame will be carried out.
+const HTTP2ReadIdleTimeout = 2 * time.Second
+
+// HTTP2PingTimeout is the timeout after which the connection will be closed  if a response to Ping is not received.
+const HTTP2PingTimeout = 2 * time.Second
+
+// HTTP2WriteByteTimeout is the timeout after which the connection will be closed no data can be written to it.
+const HTTP2WriteByteTimeout = 15 * time.Second
+
+const DNSResolveTimeout = 5 * time.Second
+
+type dependencies interface {
+	Telemetry() telemetry.Telemetry
+}
+
+// New creates new http transport intended to be used with NewSingleHostReverseProxy.
+// DNS server address is loaded from the "/etc/resolv.conf".
+func New(d dependencies, dnsServer string) (http.RoundTripper, error) {
+	return NewWithDNSServer(d, dnsServer)
+}
+
+// NewWithDNSServer creates new http transport intended to be used with NewSingleHostReverseProxy.
+func NewWithDNSServer(d dependencies, dnsServerAddress string) (http.RoundTripper, error) {
+	dialer := newDialer()
+
+	var dnsClient *dns.Client
+	var err error
+	if dnsServerAddress == "" {
+		dnsClient, err = dns.NewClientFromEtc(dialer)
+		if err != nil {
+			return nil, err
+		}
+	} else {
+		dnsClient = dns.NewClient(dialer, dnsServerAddress)
+	}
+
+	dialContext := func(ctx context.Context, network, address string) (net.Conn, error) {
+		host, port, err := net.SplitHostPort(address)
+		if err != nil {
+			return nil, err
+		}
+
+		if net.ParseIP(host) != nil {
+			return dialer.DialContext(ctx, network, address)
+		}
+
+		// Get tracing hooks
+		trace := httptrace.ContextClientTrace(ctx)
+
+		// Trace DNSStart
+		if trace != nil && trace.DNSStart != nil {
+			trace.DNSStart(httptrace.DNSStartInfo{Host: host})
+		}
+
+		// Create context for DNS resolving
+		// It separates the events/tracing of the connection to the DNS server, from the connection to the target server.
+		resolveCtx, cancel := context.WithTimeout(context.Background(), DNSResolveTimeout)
+		defer cancel()
+
+		// We are using custom DNS resolving to detect if the target host - data app, is running.
+		// The DNS query is not recursive and not cached, see "dns" package for details.
+		ip, err := dnsClient.Resolve(resolveCtx, host) //nolint:contextcheck
+
+		// Trace DNSDone
+		if trace != nil && trace.DNSDone != nil {
+			trace.DNSDone(httptrace.DNSDoneInfo{
+				Addrs: []net.IPAddr{{IP: net.ParseIP(ip)}},
+				Err:   err,
+			})
+		}
```

Comment:
This issue maybe was reason of unexpected `502` errors, ... because we detected the connection to the DNS server as sucessful conection to the app, .. and then it failed in Go reverse proxy. (cc @jachym-tousek-keboola @hosekpeter)

But maybe there are some other issues, we will see after manual testing in the testing stack.